src/main/flight/alt_hold.c

```diff
@@ -0,0 +1,274 @@
+    // user throttle must be not more than half way out from hover for a stable hold
+}
+
+void altHoldUpdateTargetAltitude(void)
+{
+    // this was part of the oritinal code, modifed a bit to be either side of hover point.
+    // it works very well.
+    // The user van raise or lower the target altitude with throttle up;  there is a big deadband
+    // Max rate for climb/descend is 1m/s by default (up to 2.5 is allowed but overshoots a fair bit)
+
+    // some people may not like making the target altitude adustable, because 
+    // with throttle adjustment, hitting the switch won't always hold altitude if throttle is bumped
+    // eg if the throttle is bumped low accidentally, quad will start descending.
+    // on the plus side, the pilot has control nice control over altitude, and the deadband is wide
+    // if the craft is stable, the throttle can't be full up or full down
+    // this is helpful when exiting the hold, avoiding bad sudden climbs or falls
+
+    const float rcThrottle = rcCommand[THROTTLE];
+
+    const float lowThreshold = 0.5f * (altholdConfig()->altHoldThrottleHover + PWM_RANGE_MIN); // around 1150
+    const float highThreshold = 0.5f * (altholdConfig()->altHoldThrottleHover + PWM_RANGE_MAX); // around 1500
+    
+    float scaler = 0.0f;
+    if (rcThrottle < lowThreshold) {
+        scaler = scaleRangef(rcThrottle, PWM_RANGE_MIN, lowThreshold, -1.0f, 0.0f);
+    } else if (rcThrottle > highThreshold) {
+        scaler = scaleRangef(rcThrottle, highThreshold, PWM_RANGE_MAX, 0.0f, 1.0f);
+    }
+
+    const float altitudeTargetDelta = scaler * altholdConfig()->altHoldTargetAdjustRate * 0.01f;
+    // task is 100Hz, default adjustRate of 100 adds/subtracts 1m every second at full stick position
+    const float newTargetAltitude = altHoldState.targetAltitudeCm + altitudeTargetDelta;
+    altHoldState.targetAltitudeCm = newTargetAltitude;
+}
+
+void altHoldUpdate(void)
+{
+    altHoldProcessTransitions(); // and initialise values if we start
+
+    // things that always need to happen in the background should go here
+    altHoldState.measuredAltitudeCm = getAltitude();
+    DEBUG_SET(DEBUG_ALTHOLD, 1, (lrintf)(altHoldState.measuredAltitudeCm));
+
+    // exit unless altHold is active
+    if (!altHoldState.altHoldActive) {
+        DEBUG_SET(DEBUG_ALTHOLD, 0, 0);
+        return;
+    }
+
+
+    // ** for testing purposes... **
+
+    // estimate the vertical velocity from vertical accelerometer
+    // not used at present, but logged.
+    // my testing doesn't support using Accelerometer data for altitude control
+    // simple Z axis value might be OK when altHold is active mode since, unless windy, quad is close to level
+    t_fp_vector accelerationVector = {{
+        acc.accADC[X],
+        acc.accADC[Y],
+        acc.accADC[Z]
+    }};
+    imuTransformVectorBodyToEarth(&accelerationVector);
+    float measuredAccel = 9.8f * (accelerationVector.V.Z - acc.dev.acc_1G) / acc.dev.acc_1G; // m/s^2
```

Comment:
To be honest I don't understand exactly what the `.V` was supposed to achieve there?

Reply:
it will work only if you'll use fpVector3_t instead of u_fp_vector. V comes from u_fp_vector definition as internal structure with t_fp_vector_def type
```
typedef union u_fp_vector {
    float A[3];
    t_fp_vector_def V;
} t_fp_vector;
```

Reply:
See comment below...